qabelbox/src/main/java/de/qabel/qabelbox/box/interactor/BoxFileBrowser.kt

```diff
@@ -0,0 +1,176 @@
+                return@thread
+            }
+            val entry = boxObject.toEntry()
+            if (entry  == null) {
+                subscriber.onError(FileNotFoundException("File or Folder ${path.name} not found"))
+                return@thread
+            }
+            subscriber.onNext(entry)
+        }
+    }
+
+    private fun queryObject(path: BoxPath): BoxObject =
+           queryObjectAndNav(path).first
+
+    private fun queryObjectAndNav(path: BoxPath): Pair<BoxObject, BoxNavigation> {
+        with(navigateTo(path.parent)) {
+            return Pair(listFiles().find { it.name == path.name } ?:
+                    listFolders().find { it.name == path.name } ?:
+                    throw FileNotFoundException("Not found: ${path.name}"),
+                    this)
+        }
+    }
+
+    override fun upload(path: BoxPath.File, source: UploadSource): Observable<Unit> = observable {
+        thread {
+            recursiveCreateFolder(path.parent).upload(path.name, source.source, source.entry.size)
+            it.onNext(Unit)
+            it.onCompleted()
+        }
+    }
+
+    override fun download(path: BoxPath.File): Observable<DownloadSource> =
+        observable { subscriber ->
+            thread {
+                query(path).subscribe { entry ->
+                    if (entry is BrowserEntry.File) {
+                        subscriber.onNext(DownloadSource(entry, navigateTo(path.parent).download(path.name)))
+                    } else {
+                        subscriber.onError(FileNotFoundException("Not a file"))
+                    }
+                }
+            }
+    }
+
+    override fun delete(path: BoxPath): Observable<Unit> = observable {
+        subscriber ->
+        thread {
+            try {
+                val nav = navigateTo(path.parent)
+                when (path) {
+                    is BoxPath.Folder -> nav.getFolder(path.name).let { nav.delete(it) }
+                    is BoxPath.File -> nav.getFile(path.name).let { nav.delete(it) }
+                }
+            } catch (e: QblStorageNotFound) {
+            } catch (e: QblStorageException) {
+                subscriber.onError(e)
+                return@thread
+            }
+            subscriber.onNext(Unit)
+        }
+    }
+
+    override fun list(path: BoxPath.FolderLike): Observable<List<BrowserEntry>> = observable {
+        subscriber ->
+        thread {
+            val nav = try {
+                navigateTo(path)
+            } catch (e: QblStorageException) {
+                subscriber.onError(e)
+                return@thread
+            }
+            nav.reloadMetadata()
+            val entries = nav.listFiles() + nav.listFolders()
+            subscriber.onNext(entries.map { it.toEntry() }.filterNotNull())
+        }
+    }
+
+    override fun createFolder(path: BoxPath.FolderLike): Observable<Unit> =
+        observable { subscriber ->
+            thread {
+                if (path !is BoxPath.Root) {
```

Comment:
what happens if this `if` wasn't here? Can't find a test for it and I guess except for an unneccessary `subscriber.onNext(Unit)`, nothing would happen because navigateTo won't execute actions on root